ls1intum/Artemis
# Consistent date formatting v2

### Checklist

### Motivation and Context
Currently, we are using more than 7 different date and time formats. Additionally, [Angular's date pipe](https://angular.io/api/common/DatePipe), that is used to format dates throughout the project, does not apply the current user's language setting.

### Description
- [x] I've implemented a custom `ArtemisDatePipe` that basically supports a long and a short format and is aware of the current user's language setting. The pipe returns always a short format on devices with less than 768 pixel width.
- [x] I've added test cases for alle parameter combinations of the `ArtemisDatePipe` for the `en` and `de` locales.
- [x] I've couples all usages of moment.js `format()` with the static `format()` method of the `ArtemisDatePipe`
- [x] I've replaced all usages of Angular's date pipe with the new `ArtemisDatePipe`.

**Why have I implemented a custom `ArtemisDatePipe`?**
- [Angular's Date Pipe](https://github.com/angular/angular/blob/master/packages/common/src/pipes/date_pipe.ts) cannot be extended to apply the current user's language setting as it is a stateless (`pure: true`) pipe.
- In contrast to the formats available with Angular's date pipe, moment.js provides localized date and time formats.
- Single point for changes in the future.
- Pleasant side-effect: the IDE recognizes the import of the custom pipe.

**Date formats**
**_format_**|**de**|**en**
:-----:|:-----:|:-----:
**short**|17.12.19 0:43|17/12/19 0:43
**long**|17. Dez. 2019 00:43|Dec 17, 2019 12:43 AM
**short-date**|17.12.19|17/12/19
**long-date**|17. Dez. 2019|Dec 17, 2019
**short-time**|0:43|0:43
**long-time**|00:43|12:43 AM

**Usage examples and output**
`{{ course.startDate | artemisDate:format:time:seconds }}`
- `{{ course.startDate | artemisDate }}`
   - `en` locale: `Dec 17, 2019 12:43 AM`
   - `de` locale: `17. Dez. 2019 00:43`
- `{{ course.startDate | artemisDate:'short' }}`
   - `en` locale: `17/12/19 0:43`
   - `de` locale: `17.12.19 0:43`
- `{{ course.startDate | artemisDate:'long-date' }}`
   - `en` locale: `Dec 17, 2019`
   - `de` locale: `17. Dez. 2019`
- `{{ course.startDate | artemisDate:'short-time':true }}`
   - `en` locale: `0:43:13`
   - `de` locale: `0:43:13`

**Follow-up PRs**
- Remove `ngx-moment`
- Use the moment adapter for ng-pick-datetime: https://danielykpan.github.io/date-time-picker/ (`format-date.pipe.ts` can then be removed)

### Steps for Testing
- Check all places with date strings. Does the selected format fit there?
- Especially check the components [where I've changed TypeScript files](https://github.com/ls1intum/Artemis/pull/1488/files?file-filters%5B%5D=.ts) as the changes I made there might lead to functional and not only UI-related issues

### Screenshots
**`en` locale:**
![image](https://user-images.githubusercontent.com/15168372/82220850-7d184580-991f-11ea-8a4b-3783467cd96b.png)

**`de` locale:**
![image](https://user-images.githubusercontent.com/15168372/82221011-b355c500-991f-11ea-9c0f-944acb1b2bfd.png)
